<launch>
  <!-- Common parameters -->
  <arg name="camera_name" default="pointgrey" />
  <arg name="frame_rate" default="170" />
  <arg name="exposure"      value="0.0035" />
  <arg name="auto_exposure"      value="false" />
  <arg name="auto_gain"     value="false" />
  <arg name="gain"          value="10" />
  <arg name="enable_trigger"          value="true" />
  <arg name="trigger_source"          value="GPIO0" />
  <arg name="trigger_polarity"          value="0" />
  <arg name="gamma"          value="0.6" />

  <arg name="camera_serial_l" default="16276501" />
  <arg name="camera_calibrated_l" default="true" />

  <arg name="camera_serial_r" default="16276527" />
  <arg name="camera_calibrated_r" default="true" />
  <arg name="nodelet_manager_name" value="camera_nodelet_manager" />
  <arg name="nodelet_manager_name_absolute" value="/$(arg camera_name)/$(arg nodelet_manager_name)" />
  <arg name="use_scale" value="true" />
  <arg name="scale_width" value="0.5" />
  <arg name="scale_height" value="0.5" />
  <arg name="interpolation" value="linear"/>
  <!--arg name="width" value="640" /-->
  <!--arg name="height" value="512" /-->

  <group ns="$(arg camera_name)" >
    <node name="$(arg nodelet_manager_name)" pkg="nodelet" type="nodelet" args="manager" />

    <!-- Both cameras are not loaded into one nodelet manager to avoid the namespacing issue. -->
    <node name="pointgrey_nodelet" pkg="nodelet" type="nodelet"
          args="load pointgrey_camera_driver/PointGreyStereoCameraSPNodelet $(arg nodelet_manager_name)">
      <param name="frame_id_l" value="camera_l" />
      <param name="frame_id_r" value="camera_r" />
      <param name="serial_l" value="$(arg camera_serial_l)" />
      <param name="serial_r" value="$(arg camera_serial_r)" />
      <param name="camera_id_l" value="0" />
      <param name="camera_id_r" value="1" />

      <!-- When unspecified, the driver will use the default framerate as given by the
           camera itself. Use this parameter to override that value for cameras capable of
           other framerates. -->
      <param name="frame_rate" value="$(arg frame_rate)" />
      <param name="auto_exposure" value="$(arg auto_exposure)" />
      <param name="exposure" value="$(arg exposure)" />
      <param name="gamma" value="$(arg gamma)" />
      <param name="auto_gain" value="$(arg auto_gain)" />
      <param name="gain" value="$(arg gain)" />
      <param name="enable_trigger" value="$(arg enable_trigger)" />
      <param name="trigger_source" value="$(arg trigger_source)" />
      <param name="trigger_polarity" value="$(arg trigger_polarity)" />

      <!-- Use the camera_calibration package to create this file -->
      <param name="camera_info_url_l" if="$(arg camera_calibrated_l)"
             value="file://$(env HOME)/.ros/camera_info/$(arg camera_serial_l).yaml" />
      <param name="camera_info_url_r" if="$(arg camera_calibrated_r)"
             value="file://$(env HOME)/.ros/camera_info/$(arg camera_serial_r).yaml" />
      <!--remap from="left/image_raw" to="left/image_raw_large" />
      <remap from="right/image_raw" to="right/image_raw_large" />
      <remap from="left/camera_info" to="left/camera_info_large" />
      <remap from="right/camera_info" to="right/camera_info_large" /-->
    </node>

    <include file="$(find image_proc)/launch/image_proc.launch" ns="left">
        <arg name="manager" value="$(arg nodelet_manager_name_absolute)" />
    </include>
    <group ns="left">
        <node name="hsv" pkg="nodelet" type="nodelet"
            args="load opencv_apps/hsv_color_filter $(arg nodelet_manager_name_absolute)">
            <remap from="image" to="image_rect_color" />
            <remap from="hsv/image" to="image_hsv" />
            <param name="h_limit_max" value="13" />
            <param name="h_limit_min" value="0" />
            <param name="s_limit_max" value="196" />
            <param name="s_limit_min" value="140" />
            <param name="v_limit_max" value="255" />
            <param name="v_limit_min" value="127" />
            <param name="debug_view" value="false" />
        </node>
        <node name="erode" pkg="nodelet" type="nodelet"
            args="load jsk_perception/ErodeMaskImage $(arg nodelet_manager_name_absolute)">
            <remap from="~input" to="image_hsv" />
            <remap from="erode/output" to="image_erode" />
            <rosparam>
                method: 0
                size: 5
                iterations: 1
            </rosparam>
        </node>
        <node name="dilate" pkg="nodelet" type="nodelet"
            args="load jsk_perception/DilateMaskImage $(arg nodelet_manager_name_absolute)">
            <remap from="~input" to="image_erode" />
            <remap from="dilate/output" to="image_dilate" />
            <rosparam>
                method: 0
                size: 12
                iterations: 2
            </rosparam>
        </node>
        <node name="centroid" pkg="nodelet" type="nodelet"
            args="load opencv_apps/contour_moments $(arg nodelet_manager_name_absolute)">
            <remap from="image" to="image_dilate" />
            <param name="debug_view" value="false" />
            <param name="use_camera_info" value="false" />
        </node>
    </group>
  </group>
</launch>
